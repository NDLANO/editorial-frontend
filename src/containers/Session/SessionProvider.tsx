/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Auth0DecodedHash } from "auth0-js";
import { createContext, Dispatch, ReactNode, SetStateAction, useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import {
  clearAccessTokenFromLocalStorage,
  getAccessToken,
  getAccessTokenPersonal,
  personalAuthLogout,
  setAccessTokenInLocalStorage,
} from "../../util/authHelpers";
import handleError from "../../util/handleError";
import { decodeToken, isValid } from "../../util/jwtHelper";
import { toLogin } from "../../util/routeHelpers";

const SessionContext = createContext<[SessionState, Dispatch<SetStateAction<SessionState>>] | undefined>(undefined);

interface Props {
  children: ReactNode;
  initialValue?: SessionState;
}

interface UserData {
  name?: string;
  permissions?: string[];
  ndlaId?: string;
}

interface SessionState {
  user: UserData;
  authenticated: boolean;
  userNotRegistered: boolean;
}

export const initialState: SessionState = {
  user: {},
  authenticated: false,
  userNotRegistered: true,
};

export interface SessionProps {
  userName?: string;
  ndlaId: string | undefined;
  userPermissions?: string[];
  authenticated: boolean;
  userNotRegistered: boolean;
  login: (authResult: Auth0DecodedHash) => void;
  logout: (federated: boolean, returnToLogin?: boolean) => void;
}

export const getSessionStateFromLocalStorage = (): SessionState => {
  const token = getAccessToken();
  const isAccessTokenPersonal = getAccessTokenPersonal();
  if (isValid(token) && isAccessTokenPersonal) {
    const decodedToken = decodeToken(token);
    return {
      user: {
        name: decodedToken?.["https://ndla.no/user_name"],
        ndlaId: decodedToken?.["https://ndla.no/ndla_id"],
        permissions: decodedToken?.permissions,
      },
      authenticated: true,
      userNotRegistered: false,
    };
  }
  return initialState; // Return initial state if token is undefined
};

export const SessionProvider = ({ children, initialValue = initialState }: Props) => {
  const sessionState = useState<SessionState>(initialValue);
  return <SessionContext value={sessionState}>{children}</SessionContext>;
};

export const useSession = (): SessionProps => {
  const sessionContext = useContext(SessionContext);
  if (sessionContext === undefined) {
    throw new Error("useSession must be used within a SessionProvider");
  }
  const [session, setSession] = sessionContext;
  const navigate = useNavigate();

  const setAuthenticated = (authenticated: boolean) => setSession((s) => ({ ...s, authenticated }));
  const setUserNotRegistered = (userNotRegistered: boolean) => setSession((s) => ({ ...s, userNotRegistered }));
  const setUserData = (user: UserData) => setSession((s) => ({ ...s, user }));

  const login = (authResult: Auth0DecodedHash) => {
    try {
      const decoded = isValid(authResult.accessToken ?? null) ? decodeToken(authResult.accessToken!) : undefined;
      const permissions = decoded?.permissions ?? [];
      const scope = decoded?.scope ?? "";
      const combinedPermissions = [...permissions, ...scope.split(" ")];
      const uniquePermissions = [...new Set(combinedPermissions)];

      if (decoded && uniquePermissions.some((permission) => permission.includes(":"))) {
        if (authResult.state) window.location.href = authResult.state;
        setAuthenticated(true);
        setUserData({
          name: decoded["https://ndla.no/user_name"],
          permissions: uniquePermissions,
        });
        setAccessTokenInLocalStorage(authResult.accessToken!, true);
        navigate("/", { replace: true });
      } else {
        setUserNotRegistered(true);
        navigate(`${toLogin()}/failure`, { replace: true });
      }
    } catch (e) {
      handleError(e);
    }
  };

  const logout = (federated: boolean, returnToLogin = false) => {
    try {
      setAuthenticated(false);
      setUserData({});
      personalAuthLogout(federated, returnToLogin);
      clearAccessTokenFromLocalStorage();
    } catch (e) {
      handleError(e);
    }
  };

  return {
    userName: session.user.name,
    ndlaId: session.user.ndlaId,
    userPermissions: session.user.permissions,
    authenticated: !!session.authenticated,
    userNotRegistered: !!session.userNotRegistered,
    login,
    logout,
  };
};
