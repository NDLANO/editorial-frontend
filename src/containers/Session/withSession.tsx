import hoistNonReactStatics from 'hoist-non-react-statics';
import { ComponentType, ReactElement } from 'react';
import { SessionProps, useSession } from './SessionProvider';

function withSession<P>(WrappedComponent: ComponentType<P & SessionProps>): ComponentType<P> {
  const WithSession = (props: P): ReactElement<P> => {
    const session = useSession();
    //@ts-ignore
    return <WrappedComponent {...{ ...props, ...session }} />;
  };

  return hoistNonReactStatics(WithSession, WrappedComponent);
}

export default withSession;
