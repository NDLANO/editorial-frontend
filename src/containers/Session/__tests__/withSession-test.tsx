import { fireEvent, getByText, render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Component } from 'react';
import { Router } from 'react-router-dom';
import { SessionProps, SessionProvider } from '../SessionProvider';
import withSession from '../withSession';

interface BaseProps {}

type Props = BaseProps & SessionProps;

const accessToken =
  'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9FSTFNVVU0T0RrNU56TTVNekkyTXpaRE9EazFOMFl3UXpkRE1EUXlPRFZDUXpRM1FUSTBNQSJ9.eyJodHRwczovL25kbGEubm8vbmRsYV9pZCI6Im1ObXJ1N1lvU0lqUGVwT3JrTjRNTWhsVSIsImh0dHBzOi8vbmRsYS5uby91c2VyX25hbWUiOiJSdW5lIFN0b3Jsw7hwYSIsImh0dHBzOi8vbmRsYS5uby9jbGllbnRfaWQiOiJXVTBLcjRDRGtyTTB1TDl4WWVGVjRjbDlHYTF2QjNKWSIsImlzcyI6Imh0dHBzOi8vbmRsYS5ldS5hdXRoMC5jb20vIiwic3ViIjoiZ29vZ2xlLW9hdXRoMnwxMDM2OTA4Njg0ODM2ODgzNzU1MzIiLCJhdWQiOiJuZGxhX3N5c3RlbSIsImlhdCI6MTUwOTk3MDAyNywiZXhwIjoxNTA5OTc3MjI3LCJhenAiOiJXVTBLcjRDRGtyTTB1TDl4WWVGVjRjbDlHYTF2QjNKWSIsInNjb3BlIjoiIn0.suhXIGs7ILtdst0NEDOPeuelfG_JZeOXlwlFTXZEgjwBCnI26e4TfrVSluCttbSuE1BJ6FuXKCtO-CjpJD6zCGj9Z00vytiWLfLq7fMYyHDfipPjUMP05H-j5TQVToTE1wLFVKLco3W9yYBWSWCBzv6GLRWWZiuzyXPyUtF2oLv30e7yAeZBA8JK10sl0DfPKrC4B1eSxrTFtvkZ3WnxeDQpvzbvZ1tHQtnF6hoarC7h9qIJFI1W95h88Mq20Ci-5k8RMILRYc1u5H5XW-RAegxI2H0EhWpCo5T4Iglwc2mUTSuetHqnbgPrxapKohcAL_b3_Nkb49CacajRt9TrQQ';

class TestComponent extends Component<Props> {
  render() {
    const { userName, login } = this.props;

    return (
      <>
        {/*  @ts-ignore */}
        <button onClick={() => login(accessToken)}>Click</button>
        <p>{userName ?? 'test'}</p>
      </>
    );
  }
}

const TestComponentWithSession = withSession(TestComponent);

describe('withSession', () => {
  it('correctly injects and updates session state', async () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Router location={history.location} navigator={history}>
        <SessionProvider>
          <TestComponentWithSession />
        </SessionProvider>
      </Router>,
    );
    expect(container).toMatchSnapshot();
    fireEvent.click(getByText(container, 'Click'));
    expect(container).toMatchSnapshot();
  });
});
