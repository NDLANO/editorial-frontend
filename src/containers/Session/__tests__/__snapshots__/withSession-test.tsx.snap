// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSession correctly injects and updates session state 1`] = `
<div>
  <button>
    Click
  </button>
  <p>
    test
  </p>
</div>
`;

exports[`withSession correctly injects and updates session state 2`] = `
<div>
  <button>
    Click
  </button>
  <p>
    test
  </p>
</div>
`;
